plugins/techdocs-backend/src/cache/cacheMiddleware.ts

```diff
@@ -0,0 +1,139 @@
+>;
+
+type CacheMiddlewareOptions = {
+  cache: TechDocsCache;
+  logger: Logger;
+};
+
+type ErrorCallback = (err?: Error) => void;
+
+export const createCacheMiddleware = ({
+  cache,
+  logger,
+}: CacheMiddlewareOptions): Router => {
+  const cacheMiddleware = router();
+
+  // And endpoint for handling cache invalidation external to the Backstage
+  // Backend (e.g. from the TechDocs CLI).
+  cacheMiddleware.use(json());
+  cacheMiddleware.post(
+    '/cache/invalidate',
+    async (req: CacheClearRequest, res) => {
+      if (req.body?.objects?.length) {
+        logger.debug(
+          `Clearing ${req.body.objects.length} cache entries: (eg: ${req.body.objects[0]})`,
+        );
+
+        try {
+          const invalidated = await cache.invalidateMultiple(req.body.objects);
+          logger.debug(
+            `Successfully invalidated ${invalidated.length} cache entries`,
+          );
+          res.status(204).send();
+        } catch (e) {
+          if (e instanceof CacheInvalidationError) {
+            const uniqueReasons = [
+              ...new Set(e.rejections.map(r => r.reason.message)),
+            ].join(', ');
+            logger.warn(
+              `Problem invalidating ${e.rejections.length} entries: ${uniqueReasons}`,
+            );
+          }
+          res.status(500).send();
+        }
+      } else {
+        res.status(400).send();
+      }
+    },
+  );
+
+  // Middleware that, through socket monkey patching, captures responses as
+  // they're sent over /static/docs/* and caches them. Subsequent requests are
+  // loaded from cache. Cache key is the object's path (after `/static/docs/`).
+  cacheMiddleware.use(async (req, res, next) => {
+    const socket = res.socket;
+    const isCacheable = req.path.includes('/static/docs/');
+
+    // Continue early if this is non-cacheable, or there's no socket.
+    if (!isCacheable || !socket) {
+      next();
+      return;
+    }
+
+    // Make concrete references to these things.
+    const reqPath = decodeURI(req.path.match(/\/static\/docs\/(.*)$/)![1]);
+    const realEnd = socket.end.bind(socket);
+    const realWrite = socket.write.bind(socket);
+    let writeToCache = true;
+    const chunks: Buffer[] = [];
+
+    // Monkey-patch the response's socket to keep track of chunks as they are
+    // written over the wire.
+    socket.write = (
+      data: string | Uint8Array,
+      encoding?: BufferEncoding | ErrorCallback,
+      callback?: ErrorCallback,
+    ) => {
+      chunks.push(Buffer.from(data));
+      if (typeof encoding === 'function') {
+        return realWrite(data, encoding);
+      }
+      return realWrite(data, encoding, callback);
+    };
+
+    // When a socket is closed, if there were no errors and the data written
+    // over the socket should be cached, cache it as a base64-encoded string!
+    socket.on('close', hadError => {
+      if (writeToCache && !hadError) {
+        cache.set(reqPath, Buffer.concat(chunks));
+      }
+    });
```

Comment:
Just curious: is a client completely done processing the response and the cache write is happening "in the background"? https://nodejs.org/api/stream.html#stream_event_close indicates it would be the case, so what happens if an exception is thrown at this place?

Reply:
Would guess the app would crash.  However, `cache.set` is just a wrapper around `this.cache.set()` in the underlying cache client, which has a `.catch()` attached that just logs the error.  So we should be safe, here.